Add Authenticated Http Proxy Support - Issue #520

coliemcgarry opened the pull request:
Added Proxy Authorization support.

New pluggable 'proxy_authorization' policy which handles http proxy authorization. The default implementation of this includes:

- Proxy Auth header parser (extracts scheme + challenge).
- Supports NTLM+Negotiate+Basic (Did not include 'Digest' at this time)
- Handles NTLM/Negotiate tokens using a security_context policy.
- Added Unit Tests for Proxy Authorization Parsing.

The change includes an implementation of a windows security_context policy using the Windows SSPI interface. This header may be better suited to live outside the library, however as I've implemented it, I think it fits pretty well with the current design. It only comes into play if you include this policy (via config), which results (on windows in extra link time dependencies)

I added a new config policy to show how to include a proxy authenticated web socket i.e. asio_tls_client_authenticated_proxy.

I needed a small API change i.e. to handle re-connection request from the proxy. When a proxy_reconnect is reported from the library, the client code drops current connection, and creates a new connection. The life cycle of the proxy_authorization object is a little difficult, so during a reconnection, the new connection object takes ownership of the proxy_authenticator object. By doing so, we only need keep proxy authenticators with the associated connection objects. 

This pull request is an update of previous pull request #532 and #534. The first update (#534) moved ownership of the proxy authenticator object to the connection object. This second update is mainly some tidy ups and I re-based to the latest 'develop' branch.

zaphoyd commented:
Hi @coliemcgarry sorry for the delay on this. Wanted to acknowledge that i've seen this PR, am definitely interested in getting this sort of change into the library. Thanks for this work so far! As this is a sizable change I'll probably need a bit of time to fully review it.

coliemcgarry commented:
Apologies for the size of the change, and if there's a way to shrink this please let me know. 

I've tested locally on a proxy we use, and we're good, however I also tested on a customer's proxy, and I hit some issue - I'll be revisiting this in the next few days, so I may have another update!

coliemcgarry commented on websocketpp/http/impl/response.hpp:
```diff
@@ -137,20 +137,19 @@ inline size_t response::consume(char const * buf, size_t len) {
     while (s.good()) {
-        s.getline(buf,istream_buffer);
-        bytes_read = static_cast<size_t>(s.gcount());
+        std::string line;
```
I needed to rework this method to handle long responses from proxies. 
Some proxies return large HTML blobs along with the '407', so we need to sink this data before moving on with authorization.

coliemcgarry commented:
So I addressed the issues seen at a customers site:
1) Case sensitivity issues when comparing the auth scheme name.
2) Proxy returns large HTML blob along with the 407, and we need to sink this data before carrying on with proxy authorization.

coliemcgarry commented:
I updated my pull request again. I rebased against the latest code in 'develop'. 

I needed to update to address a customer found issue, where a proxy behaves differntly. 

Please see the updated pull request: https://github.com/zaphoyd/websocketpp/pull/567
